DistroLauncher/named_mutex.h

```diff
@@ -0,0 +1,219 @@
+ *  ```
+ *
+ *  In order to acquire a lock, you have two options.
+ *  Use the first one unless you've got a good reason not to.
+ *   1. Use the monadic interface:
+ *      ```
+ *      my_mutex.lock().and_then([](){
+ *          // do stuff
+ *      }).or_else([](){
+ *          throw "Oh, no!";
+ *      });
+ *      ```
+ *      Note that the ordering of the functions doesn't matter,
+ *      and both are optional and repeatable, so long as no
+ *      exceptions are thrown. In this case, the mutex will be
+ *      released and the exception re-thrown.
+ *
+ *   2. Use a scope guard:
+ *      ```
+ *      NamedMutex::Guard scope_guard = my_mutex.lock();
+ *      if(scope_guard.ok()) {
+ *          // do stuff
+ *      } else {
+ *          throw "Oh, no!";
+ *      }
+ *      ```
+ *      Be careful! If your program throws and it's not caught,
+ *      your mutex may not be released. Option 1 takes care of
+ *      this automatically, hence why it's recommended.
+ */
+
+// Templated for testing/dependency injection reasons, always use the alias at the end of the file
+template <typename MutexAPI> class NamedMutexWrapper
+{
+  public:
+    NamedMutexWrapper() = delete;
+    NamedMutexWrapper(NamedMutexWrapper&) = delete;
+    NamedMutexWrapper(NamedMutexWrapper&&) = delete;
+
+    /// Some mutexes are almost never used. This makes for an easy optimization by not initializing them unless needed.
+    NamedMutexWrapper(std::wstring name, bool lazy_init = false) :
+        mutex_handle(nullptr), mutex_name(L"WSL_" + DistributionInfo::Name + L"_" + name)
+    {
+        if (!lazy_init) {
+            create();
+        }
+    }
+
+    virtual ~NamedMutexWrapper() noexcept
+    {
+        destroy();
+    }
+
+    class Lock
+    {
+      public:
+        constexpr Lock() noexcept : parent_(nullptr), response_(0)
+        {
+        }
+
+        Lock(NamedMutexWrapper& parent) noexcept : parent_(&parent), response_(parent.wait_and_acquire())
+        {
+        }
+
+        Lock(Lock& other) = delete;
+        Lock(Lock&& other) noexcept : Lock()
+        {
+            *this = std::move(other);
+        }
+
+        ~Lock() noexcept
+        {
+            release();
+        }
+
+        void release() noexcept
+        {
+            if (ok()) {
+                parent_->release();
+            }
+            parent_ = nullptr;
+        }
+
+        Lock& operator=(Lock&& other) noexcept
+        {
+            if (this != &other) {
+                parent_ = std::exchange(other.parent_, nullptr);
+                response_ = std::exchange(other.response_, 0);
+            }
+            return *this;
+        }
+
+        bool ok() const noexcept
+        {
+            return parent_ && (response_ == WAIT_OBJECT_0);
+        }
+
+        DWORD why() const noexcept
+        {
+            return parent_ ? 0 : response_;
+        }
+
+        template <typename Callable> Lock&& and_then(Callable&& f)
+        {
```

Comment:
Are you sure we need to make the return types rvalue references? This seems to create complexity that may not be necessary.

Reply:
Yes, otherwise this doesn't compile:
```
mutext.lock().and_then(function).or_else(other_function);
             ^                  ^ 
Deleted constructor    Deleted constructor
```

Reply:
I though rvalue return optimization would save the day here, but apparently not, so I had to do this 🤷.

Reply:
Yep, you also have to delete the copy-constructor:

https://godbolt.org/z/1q7EqoEWz

Reply:
You're right, seems like I misread the compiler at some point.

Reply:
I just found what it actually broke:
```c++
auto scope_lock = mutex.lock().and_then([] { puts("Hello!"); });
                ^ Deleted constructor Lock(Lock&)
```
Honestly I'm fine breaking this, it further discourages the scope-lock idiom. Ultimately, it can still be done with:
```c++
auto scope_lock = mutex.lock();
scope_lock.and_then([] { puts("Hello!"); });
```